aggregator/src/datastore.rs

```diff
@@ -1344,7 +1315,8 @@ impl<C: Clock> Transaction<'_, C> {
         let stmt = self
             .prepare_cached(
                 "INSERT INTO client_reports (task_id, report_id, client_timestamp)
-                VALUES ((SELECT id FROM tasks WHERE task_id = $1), $2, $3)",
+                VALUES ((SELECT id FROM tasks WHERE task_id = $1), $2, $3)
+                ON CONFLICT DO NOTHING",
```

Comment:
FWIW: in the presence of a misbehaving leader on a task with a VDAF allowing multiple aggregation parameters, `ON CONFLICT DO NOTHING` could lead to a surprise. If the leader incorrectly sends two different reports sharing the same report ID, in aggregation jobs with different aggregation parameters, then the helper will save the first report's timestamp, and eventually reuse that timestamp when preparation of the second report finishes. This timestamp could end up pointing to the "wrong" batch aggregation, compared to the actual timestamp on the second report, and result in two batch aggregations being out of sync between the leader and helper.

The above describes a robustness failure in the presence of a misbehaving aggregator, which is acceptable in the security model, but we may want to catch the changing timestamp in service of debugging the other aggregator. I think the following would let us catch such issues by throwing an error if no rows were modified.

```suggestion
                ON CONFLICT (task_id, report_id) DO UPDATE SET client_timestamp = $3 WHERE excluded.client_timestamp = $3",
```

(or, we could use a PL/pgSQL function to do a similar check in a single DB round-trip, but avoid unnecessary tuple writes)

Reply:
What you describe is a regression introduced by this change (we used to reject aggregation jobs if any report ID therein already existed) so I want to fix it here. I think the behavior we want is to raise an error if we detect an attempt to mutate an existing `client_reports` row. I figured out a way to do this using the `xmax` system column (actually I'm cribbing from StackOverflow, naturally). I'd appreciate some feedback on whether this is too complicated.

Reply:
I'm wary about using MVCC metadata, particularly given the caveats about `xmax` being nonzero due to rollbacks or locking.

I fleshed out my suggestion snippet above into 7f4167b, this passes the new tests, and has a downside of performing unneccessary no-op UPDATEs on the Poplar1 happy path. If we want to avoid extra UPDATEs, I'd prefer we do so via a PL/pgSQL function, so that we can SELECT, conditionally INSERT, and return a flag all in one call.

Reply:
Fair enough. I suspect we have a bunch more work to do before VDAFs with aggregation parameters actually work in Janus, so I'm happy to take a Poplar1 performance hit for now. Thank you for looking into this.

Reply:
I like the current `UPDATE` trick to detect if anything is actually changing. This seems more understandable than a trick based on system columns, and we definitely don't need to write a pgSQL function to avoid in-DB no-op updates (yet).